File: CPlusPlus/42. 接雨水.cpp
/*
某个节点能够接下的雨水就是 min（左边最高，右边最高）- 节点值
*/
class Solution {
public:
    int trap(vector<int>& height) {
        int sz = height.size();
        vector<int> leftmax(sz,0);
        vector<int> rightmax(sz,0);
        leftmax[0] = height[0];
        int current_max = leftmax[0];
        for(int i = 1; i < height.size(); ++i){
            if(height[i] > current_max){
                current_max = height[i];
            }
            leftmax[i] = current_max;
        }
        rightmax[height.size()-1] = height[height.size()-1];
        current_max = rightmax[height.size()-1];
        for(int i = height.size()-1; i >=0; --i){
            if(height[i] > current_max){
                current_max = height[i];
            }
            rightmax[i] = current_max;
        }
        int sum = 0;
        for(int i = 0; i < height.size(); ++i){
            sum += std::min(leftmax[i],rightmax[i]) - height[i];
        }
        return sum;
        
    }  
};

File: Rust/42. 接雨水.rs
// 动态规划， 记录左边&右边的最大值
use std::cmp;
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        let mut prefix: Vec<i32> = vec![0; n];
        let mut postfix = vec![0; n];
        prefix[0] = height[0];
        for i in 1..n {
            prefix[i] = cmp::max(prefix[i - 1], height[i]);
        }
        postfix[n - 1] = height[n - 1];
        for i in (0..(n - 1)).rev() {
            postfix[i] = cmp::max(postfix[i + 1], height[i]);
        }
        let mut res = 0i32;
        for i in 1..n - 1 {
            let curr = cmp::min(prefix[i - 1], postfix[i + 1]);
            if curr > height[i] {
                res = res + (curr - height[i]);
            }
        }
        res
    }
}
